waiter/src/waiter/auth/oidc.clj

```diff
@@ -0,0 +1,211 @@
+        authorize-uri (jwt/retrieve-authorize-url
+                        jwt-auth-server request oidc-callback-uri code-verifier state-code)]
+    (-> response
+      (assoc :status http-302-moved-temporarily)
+      (update :headers assoc
+              "cache-control" "no-store"
+              "content-security-policy" content-security-policy-value
+              "location" authorize-uri)
+      (cookie-support/add-encoded-cookie
+        password oidc-challenge-cookie code-verifier challenge-cookie-duration-secs))))
+
+(defn make-oidc-auth-response-updater
+  "Returns a response updater that rewrites 401 waiter responses to 302 redirects."
+  [jwt-auth-server password request]
+  (fn update-oidc-auth-response [{:keys [status] :as response}]
+    (if (and (= status http-401-unauthorized)
+             (utils/waiter-generated-response? response))
+      ;; issue 302 redirect
+      (trigger-authorize-redirect jwt-auth-server password request response)
+      ;; non-401 response, avoid authentication challenge
+      response)))
+
+(defn supports-redirect?
+  "Returns true when:
+   - either the accept-redirect-auth request header is present in the request
+   - or the request is deemed to have come from a browser."
+  [oidc-authority request]
+  (or (hu/browser-request? request)
+      (and (= "yes" (get-in request [:headers accept-redirect-header-name]))
+           (let [accept-redirect-auth (get-in request [:headers accept-redirect-auth-header-name])]
+             (or (str/blank? accept-redirect-auth)
+                 (= "." accept-redirect-auth)
+                 (some #(= oidc-authority %) (str/split accept-redirect-auth #" ")))))))
+
+(defn wrap-auth-handler
+  "Wraps the request handler with a handler to trigger JWT access token authentication."
+  [{:keys [allow-oidc-auth-api? allow-oidc-auth-services? jwt-auth-server oidc-authorize-uri password]} request-handler]
+  (let [oidc-authority (-> oidc-authorize-uri (URI.) (.getAuthority))]
+    (fn oidc-auth-handler [{:keys [waiter-api-call?] :as request}]
+      (let [use-oidc-auth? (or
+                             ;; service requests will enable OIDC auth based on env variable or when absent, allow-oidc-auth-services?
+                             (and (not waiter-api-call?)
+                                  (= "true" (get-in request [:waiter-discovery :service-description-template "env" "USE_OIDC_AUTH"]
+                                                    (str allow-oidc-auth-services?))))
+                             ;; waiter api requests will enable OIDC auth based on allow-oidc-auth-api?
+                             (and waiter-api-call? allow-oidc-auth-api?))
+            use-redirect? (supports-redirect? oidc-authority request)]
+        (cond
+          (or (not use-oidc-auth?)
+              (not use-redirect?) ;; OIDC auth is no-op when request cannot be redirected
+              (auth/request-authenticated? request))
```

Comment:
We have the same workflow in JWT. 
All the auth handlers (JWT, kerberos, saml) are invoked in a chain, they opt to avoid work when the request is already authenticated.

Reply:
What I mean is, we can avoid all the logic in `supports-redirect?` for example if someone is already authed.

Reply:
Restructured the sequence of checks.